flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/source/reader/TestColumnStatsWatermarkExtractor.java

```diff
@@ -0,0 +1,215 @@
+
+  private GenericAppenderHelper dataAppender;
+  private long minTs = Long.MAX_VALUE;
+  private long minTsTz = Long.MAX_VALUE;
+  private long minL = Long.MAX_VALUE;
+
+  @Before
+  public void initTable() throws IOException {
+    dataAppender =
+        new GenericAppenderHelper(
+            sourceTableResource.table(), FileFormat.PARQUET, TEMPORARY_FOLDER);
+
+    List<Record> batch = RandomGenericData.generate(SCHEMA, 3, 2L);
+    dataAppender.appendToTable(batch);
+
+    for (Record r : batch) {
+      LocalDateTime localDateTime = (LocalDateTime) r.get(0);
+      minTs = Math.min(minTs, localDateTime.toInstant(ZoneOffset.UTC).toEpochMilli());
+
+      OffsetDateTime offsetDateTime = (OffsetDateTime) r.get(1);
+      minTsTz = Math.min(minTsTz, offsetDateTime.toInstant().toEpochMilli());
+
+      minL = Math.min(minL, (Long) r.get(2));
+    }
+  }
+
+  @Test
+  public void testTimestamp() {
+    DataFile dataFile =
+        sourceTableResource
+            .table()
+            .currentSnapshot()
+            .addedDataFiles(sourceTableResource.table().io())
+            .iterator()
+            .next();
+    ColumnStatsWatermarkExtractor tsExtractor =
+        new ColumnStatsWatermarkExtractor(SCHEMA, "ts", null);
+
+    Assert.assertEquals(
+        minTs,
+        tsExtractor.extractWatermark(
+            IcebergSourceSplit.fromCombinedScanTask(new DummyTask(dataFile))));
+  }
+
+  @Test
+  public void testTimestampWithTz() {
+    DataFile dataFile =
+        sourceTableResource
+            .table()
+            .currentSnapshot()
+            .addedDataFiles(sourceTableResource.table().io())
+            .iterator()
+            .next();
+    ColumnStatsWatermarkExtractor tsTzExtractor =
+        new ColumnStatsWatermarkExtractor(SCHEMA, "tstz", null);
+
+    Assert.assertEquals(
+        minTsTz,
+        tsTzExtractor.extractWatermark(
+            IcebergSourceSplit.fromCombinedScanTask(new DummyTask(dataFile))));
+  }
+
+  @Test
+  public void testLong() {
+    DataFile dataFile =
+        sourceTableResource
+            .table()
+            .currentSnapshot()
+            .addedDataFiles(sourceTableResource.table().io())
+            .iterator()
+            .next();
+    ColumnStatsWatermarkExtractor longExtractorMilliSeconds =
+        new ColumnStatsWatermarkExtractor(SCHEMA, "l", TimeUnit.MILLISECONDS);
+    ColumnStatsWatermarkExtractor longExtractorMicroSeconds =
+        new ColumnStatsWatermarkExtractor(SCHEMA, "l", TimeUnit.MICROSECONDS);
+
+    Assert.assertEquals(
+        minL,
+        longExtractorMilliSeconds.extractWatermark(
+            IcebergSourceSplit.fromCombinedScanTask(new DummyTask(dataFile))));
+    Assert.assertEquals(
+        minL / 1000L,
+        longExtractorMicroSeconds.extractWatermark(
+            IcebergSourceSplit.fromCombinedScanTask(new DummyTask(dataFile))));
+  }
+
+  @Test
+  public void testMultipleFiles() throws IOException {
```

Comment:
following this style, maybe the previous 3 methods can be merged into one test method `testSingleFile`?

Reply:
Ryan always presses that we test every feature in its own method.